src/hiccup/compiler.clj

```diff
@@ -73,37 +73,48 @@
   (or content
       (and (html-mode?) (not (void-tags tag)))))
 
-(defn- merge-classes [class classes]
+(defn- merge-classes [class classes compiling?]
   (cond
-    (nil? class)    classes
-    (string? class) (str classes " " class)
-    :else           (str classes " " (str/join " " class))))
-
-(defn- merge-attributes [map-attrs id classes]
+    (nil? class)                  classes
+    (string? class)               (str classes " " class)
+    (and compiling?
+         (symbol? class))         (list `str classes " " class)
+    (and (not compiling?)
+         (instance? Named class)) (str classes " " (name class))
+    :else                         (str classes " " (str/join " " class))))
+
+(defn- merge-attributes [map-attrs id classes compiling?]
```

Comment:
Again, we want two separate functions here: `merge-attributes` and `merge-attributes-form`.